### DISCUSSION TITLE
NETBOX Installation proccess - SECRET_KEY must be at least 50 characters in length

### DISCUSSION THREAD

**Discussion by vdodin** [netbox-community/netbox#15549] [Help Wanted!]
Hello there, 

I'm pretty new to Netbox. Trying to install it on Linux Ubuntu 22.04.3 LTS. I'm using this method : https://github.com/netbox-community/netbox/blob/develop/docs/installation/3-netbox.md 

Under the section of SECRET_KEY, I generated a key using this: python3 `/opt/netbox/netbox/netbox/generate_secret_key.py`
and put it into` /opt/netbox/netbox/netbox/netbox/configuration.py `  under SECRET_KEY, right after Redis section in that file. 

After updating the file, I run sudo /opt/netbox/upgrade.sh where I receive this at the end:
```
File "/opt/netbox/netbox/netbox/netbox/settings.py", line 76, in <module>
    raise ImproperlyConfigured(
django.core.exceptions.ImproperlyConfigured: SECRET_KEY must be at least 50 characters in length. To generate a suitable key, run the following command:
  python /opt/netbox/netbox/netbox/generate_secret_key.py
```

Tried different options to generate the key in different generators, also manually. What do I miss? 

Thanks in advance!

**Comment by candlerb** [comment]
Netbox version?

The secret key needs to be 50 characters in length, and the generated one should be fine (+1 for newline here):

```
# python3 /opt/netbox/netbox/generate_secret_key.py  | wc
      1       1      51
```

You might have put it in the wrong place.

It's possible you've made one too many nested "netbox" subdirectories, because my system has `/opt/netbox/netbox/generate_secret_key.py` but you said yours was `/opt/netbox/netbox/netbox/generate_secret_key.py`.

By the same token, the file which is normally at `/opt/netbox/netbox/netbox/configuration.py` may be located in a different location on your system.

If this is the problem, you'll make life easier for yourself if you stick to the default installation location of `/opt/netbox` rather than `/opt/netbox/netbox`; all the other configs and examples you see will be assuming the former.

**Reply by markkuleinio** [reply]
Just an idea: note that there is a dot at the end of the "git clone" command in the docs

**Reply by vdodin** [reply]
Hi, 
Thanks for you response! 
The Netbox version is 3.7.0 
You're right. But I'm not sure how did I do that with all these additional directories. Do you think that I should try and reinstall everything from scratch?

**Reply by candlerb** [reply]
Did you install from git? If /opt/netbox is a real directory which contains a netbox subdirectory and nothing else, then I think it should be OK to move all the sub-subdirectories up one level (including the hidden ones), then re-run `/opt/netbox/upgrade.sh` to rebuild the python virtualenv.

```
cd /opt/netbox
mv netbox tmp
mv tmp/* tmp/.??* ./
rmdir tmp
./upgrade.sh
```

And make sure that your systemd configs point to the correct locations, as per the [contrib](https://github.com/netbox-community/netbox/tree/v3.7.4/contrib) examples.

**Reply by vdodin** [reply]
The thing is, that I have 2 sub directories under opt/netbox 
```
netbox-ng: /opt/netbox/netbox$ ls
base_requirements.txt  contrib          docs         media       netbox  pyproject.toml  reports           scripts      upgrade.sh
CHANGELOG.md           CONTRIBUTING.md  LICENSE.txt  mkdocs.yml  NOTICE  README.md       requirements.txt  SECURITY.md  venv
```
and 

```
@netbox-ng:/opt/netbox/netbox/netbox$ ls
account   core  extras                  ipam       media   project-static  scripts    tenancy       users      virtualization  wireless
circuits  dcim  generate_secret_key.py  manage.py  netbox  reports         templates  translations  utilities  vpn
```
Should I move all the files from both to only opt/netbox and then run upgarde.sh?

**Reply by vdodin** [reply]
P.S Installed from the procedure on git

**Comment by candlerb** [comment]
You haven't shown what you have directly under /opt/netbox. I think you will have a single directory, netbox - I.e. /opt/netbox/netbox. If so, you can follow the procedure I showed before. 

There will be other nested 'netbox' directories beneath those, that's fine, don't move anything else. But when you're finished, make sure your configuration file is /opt/netbox/netbox/netbox/configuration.py
